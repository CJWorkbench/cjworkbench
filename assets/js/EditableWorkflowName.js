import React from 'react';
import Textarea from 'react-textarea-autosize';
import PropTypes from 'prop-types'

export default class EditableWorkflowName extends React.Component {
  constructor(props) {
    super(props);
    this.saveName = this.saveName.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.keyPress = this.keyPress.bind(this);
    this.state = {
      value: this.props.value
    }
  }

  handleChange(event) {
    this.setState({value: event.target.value});
  }

  // selects the text for editing on a click
  handleClick(event) {
    if (!this.props.isReadOnly) this.textInput.childNodes[0].select();
  }

  // Make Enter key save the text in edit field, overriding default newline
  keyPress(e) {
    if (e.key == 'Enter' ) {
      e.preventDefault();
      // Blur event will trigger save 
      // Have to target child through parent b/c TextArea cannot be directly referenced
      this.textInput.childNodes[0].blur();
    }
  }

  saveName() {
    this.props.api.setWfName(this.props.wfId, this.state.value);
  }

  render() {

    return <div 
              // Saves a reference to parent to allow targeting of imported component
              ref={(input) => {this.textInput = input;}}
              onClick={this.handleClick}
              className='editable-name-container'
            >
              {this.props.isReadOnly
                ? ( <span className='editable-title-field'>{this.props.value}</span> )
                : (
                    <Textarea
                      value={this.state.value}
                      onChange={this.handleChange}
                      onBlur={this.saveName}
                      onKeyPress={this.keyPress}
                      className='editable-title-field'
                      maxRows={1}
                    />
                  )
              }
          </div>
  }
}


EditableWorkflowName.propTypes = {
  value:      PropTypes.string,
  wfId:       PropTypes.number,
  api:        PropTypes.object.isRequired,
  isReadOnly: PropTypes.bool
};
