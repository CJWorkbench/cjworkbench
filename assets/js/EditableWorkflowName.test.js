/* globals describe, expect, it, jest */
import { shallow } from 'enzyme'
import EditableWorkflowName from './EditableWorkflowName'

describe('EditableWorkflowName', () => {
  const wrapper = (extraProps = {}) => {
    return shallow(
      <EditableWorkflowName
        value='A'
        onSubmit={jest.fn()}
        isReadOnly={false}
        {...extraProps}
      />
    )
  }

  it('renders a plain title when read-only', () => {
    const w = wrapper({ isReadOnly: true })
    expect(w.find('input')).toHaveLength(0)
  })

  it('lets the user edit the title', () => {
    const onSubmit = jest.fn()
    const w = wrapper({ onSubmit })
    w.find('input').simulate('change', { target: { value: 'B' } })
    w.find('input').simulate('blur')
    expect(onSubmit).toHaveBeenCalledWith('B')
  })
})
