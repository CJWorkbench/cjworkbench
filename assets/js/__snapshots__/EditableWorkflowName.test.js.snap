// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableWorkflowName NOT Read-only Renders a title that can be edited and saved 1`] = `
<EditableWorkflowName
  api={
    Object {
      "setWfName": [Function],
    }
  }
  editClass=""
  isReadOnly={false}
  value="Test Title"
  wfId={808}
>
  <div
    onKeyPress={[Function]}
  >
    <RIETextArea
      change={[Function]}
      classEditing="editable-title-field-active"
      className=""
      propName="value"
      rows={1}
      value="Test Title"
    >
      <span
        className=""
        onClick={[Function]}
        onFocus={[Function]}
        tabIndex="0"
      >
        <span>
          Test Title
        </span>
      </span>
    </RIETextArea>
  </div>
</EditableWorkflowName>
`;

exports[`EditableWorkflowName NOT Read-only Renders a title that can be edited and saved 2`] = `
<EditableWorkflowName
  api={
    Object {
      "setWfName": [Function],
    }
  }
  editClass=""
  isReadOnly={false}
  value="Test Title"
  wfId={808}
>
  <div
    onKeyPress={[Function]}
  >
    <RIETextArea
      change={[Function]}
      classEditing="editable-title-field-active"
      className=""
      propName="value"
      rows={1}
      value="Test Title"
    >
      <textarea
        className="editable-title-field-active"
        defaultValue="Test Title"
        disabled={false}
        onBlur={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        rows={1}
      />
    </RIETextArea>
  </div>
</EditableWorkflowName>
`;

exports[`EditableWorkflowName Read-only Renders plain title 1`] = `
<div
  onKeyPress={[Function]}
>
  <span
    className=""
  >
    Test Title
  </span>
</div>
`;
